File: algorithms/Codility-master/Codility-master/7_StacksQueues/Fish.py
# 100%
def solution(A, B):
    # write your code in Python 3.6
    num = len(A)
    if num < 2: return num
    
    alive_fish = []
    
    for i in range(num):
        challenger = i
        
        while alive_fish and B[alive_fish[-1]] == 1 and B[challenger] == 0:
            surviver = alive_fish.pop()
            if A[surviver] > A[challenger]:
                challenger = surviver
                
        alive_fish.append(challenger)            
      
    return len(alive_fish)

File: algorithms/Codility-master/Codility-master/7_StacksQueues/Fish.js
// 100%
function solution(A, B) {
    // write your code in JavaScript (Node.js 8.9.4)
    
    let num = A.length;
    if (num<2) return num;
    let alive_fish = [];
    
    for (let i=0; i<num; i++){
        let challenger = i;
        
        while(alive_fish && (B[alive_fish[alive_fish.length -1]] == 1) && B[challenger] == 0){
            let surviver = alive_fish.pop();
            if(A[surviver] > A[challenger]){
                challenger = surviver;
            }
        }
        alive_fish.push(challenger);
    }
    return alive_fish.length;
}
